docs/versioning.md

```diff
@@ -0,0 +1,101 @@
+
+    // This field MUST be present for this version of the IR.
+
+By way of example, the ModelProto.ir_version MUST be present in every model.
+
+ISSUE: decide and document how we want to handle changes to the format that are forward compatible but will cause the generated code to introduce a breaking change (e.g., changing the data type of a field from int64 to int32, int64 to double or bytes to string).  I would recommend that we never do it after we hit 1.0.0 - prior to that we should be fairly liberal with changes to size (e.g., int32<>int16, float<>double) and conservative with changes to value space (e.g., integral<>floatingpoint, string<>int, scalar<>message)
+
+ISSUE: define type compatiblity rules either here or under model versioning - probably here
+
+## Operator versioning
+
+ISSUE: sort out the design and then document.
+
+## Model versioning
+
+Model authors and applications/systems  MAY elect to ignore model versioning mechanism and policy rules. For models that will be shared across developers, teams, or organizations, model authors and applications/systems SHOUD adhere to the following version policies:
+
+ISSUE: the following is a strawman. I'm confident some of it is right and some is wrong. Either way, we need to make some calls and document it.
+
+### Signature Changes
+1. Breaking changes to the ModelProto.graph.GraphProto.input our .output MUST increment the MAJOR version of ModelProto.model_version. Breaking changes include:
+    
+    * Breaking changes to the semantics of an input or output (e.g., changing the required contents of an from color image to black and white image). 
+    * Changing the declared type of an input or output to an incompatible type (e.g., tensor(int)->tensor(string).
+    * Adding a new input for which there is no meaningful or specified default value.
+    * Removing an exising output for which there is no meaningful or specified default value.
+
+2. Non-breaking changes to the ModelProto.graph.GraphProto.input our .output MUST increment the MINOR version of ModelProto.model_version. Non-breaking changes include:
+    
+    * Changing the declared type of an input or output to an compatible/widening type (e.g., tensor(int32)->tensor(int64), tensor(float16)->tensor(float32).
+    * Adding a new input for which there is no meaningful or specified default value.
+    * Removing an exising output for which there is no meaningful or specified default value.
+
+### IR version/Operator version dependency changes
+
+ISSUE: what's our policy when a model takes a dependency on new IR_VERSION change and/or new operator change?
+
+
+### Accuracy or performance changes
+
+Assuming that there are no breaking changes to the signature of the model's graph or any operator dependencies, the shape and contents of the graph can change freely provided there are no semantic changes to the model. However, changes to the shape and contents of the graph can impact model accuracy and/or model performance.    
```

Comment:
Again, this section makes sense only if we standartize particular models as part of the onnx repo (i.e. being more serious of model zoo). Is it what you have in mind?

Reply:
See comments above. Same idea.

Reply:
@dzhulgakov A good analogy is to imagine we were building an actual programming language, and are considering to add a package ecosystem. It is best if the designers of package ecosystem specify a normative specification on model definitions.